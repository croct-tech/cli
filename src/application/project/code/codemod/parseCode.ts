import {Node} from '@babel/types';
import generate from '@babel/generator';
import {Codemod, CodemodOptions, ResultCode} from '@/application/project/code/codemod/codemod';
import {Language, parse} from '@/application/project/code/codemod/parser';

export type Configuration<O extends CodemodOptions> = {
    codemod: Codemod<Node, O>,
    languages: Language[],
};

export class ParseCode<O extends CodemodOptions> implements Codemod<string, O> {
    private readonly codemod: Codemod<Node, O>;

    private readonly languages: Language[];

    public constructor(configuration: Configuration<O>) {
        this.codemod = configuration.codemod;
        this.languages = configuration.languages;
    }

    public async apply(input: string, options?: O): Promise<ResultCode<string>> {
        const result = await this.codemod.apply(parse(input, this.languages), options);

        if (!result.modified) {
            return {
                modified: false,
                result: input,
            };
        }

        return {
            modified: true,
            result: generate(
                result.result,
                {
                    compact: false,
                    retainFunctionParens: true,
                },
            ).code,
        };
    }
}
