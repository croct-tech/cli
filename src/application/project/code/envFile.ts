import {FileSystem} from '@/application/fs/fileSystem';

export class EnvFile {
    private readonly fileSystem: FileSystem;

    private readonly path: string;

    public constructor(fileSystem: FileSystem, path: string) {
        this.fileSystem = fileSystem;
        this.path = path;
    }

    public getName(): string {
        return this.fileSystem.getBaseName(this.path);
    }

    public exists(): Promise<boolean> {
        return this.fileSystem.exists(this.path);
    }

    public async hasVariable(name: string): Promise<boolean> {
        const content = await this.read();

        if (content === '') {
            return false;
        }

        const escapedVariable = EnvFile.escapeRegex(name);
        const regex = new RegExp(`^${escapedVariable}\\s*=`, 'm');

        return regex.test(content);
    }

    public async setVariables(variables: Record<string, string>): Promise<void> {
        const content = await this.read();

        try {
            for (const [name, value] of Object.entries(variables)) {
                await this.setVariable(name, value);
            }
        } catch (error) {
            await this.write(content);

            throw error;
        }
    }

    public async setVariable(name: string, value: string): Promise<void> {
        const content = await this.read();
        const escapedName = EnvFile.escapeRegex(name);

        if (content === '') {
            return this.write(`${name}=${value}`);
        }

        const entry = `${name}=${value}`;

        const updatedContent = content.replace(
            new RegExp(
                `${escapedName}\\s*=\\s*((?!['"\`]).*$|\`(?:\\.|[^\`])*\`|'(?:\\.|[^'])*'|"(?:\\.|[^"])*")`,
                'm',
            ),
            entry,
        );

        if (updatedContent !== content || updatedContent.includes(entry)) {
            return this.write(updatedContent);
        }

        return this.write(`${content}${content.endsWith('\n') ? '' : '\n'}${name}=${value}`);
    }

    private async write(content: string): Promise<void> {
        await this.fileSystem.writeTextFile(this.path, content, {
            overwrite: true,
        });
    }

    private async read(): Promise<string> {
        if (!await this.exists()) {
            return '';
        }

        return this.fileSystem.readTextFile(this.path);
    }

    private static escapeRegex(value: string): string {
        return value.replace(/[-/\\^$*+?.()|[\]{}]/g, '\\$&');
    }
}
