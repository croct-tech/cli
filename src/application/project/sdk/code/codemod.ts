export type ResultCode<T> = {
    modified: boolean,
    result: T,
};

export type CodemodOptions = Record<string, any>;

export interface Codemod<I, O extends CodemodOptions = CodemodOptions> {
    apply(input: I, options?: O): Promise<ResultCode<I>>;
}

export class CodemodError extends Error {
    public constructor(message: string) {
        super(message);

        Object.setPrototypeOf(this, CodemodError.prototype);
    }
}

export class MalformedCodeError extends CodemodError {
    public constructor(message: string) {
        super(message);

        Object.setPrototypeOf(this, MalformedCodeError.prototype);
    }
}
