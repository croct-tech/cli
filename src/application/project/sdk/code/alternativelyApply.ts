import {Codemod, CodemodOptions, ResultCode} from '@/application/project/sdk/code/codemod';

export class AlternativelyApply<T, O extends CodemodOptions> implements Codemod<T, O> {
    private readonly codemods: Array<Codemod<T, O>>;

    public constructor(codemod: Codemod<T, O>, ...codemods: Array<Codemod<T, O>>) {
        this.codemods = [codemod, ...codemods];
    }

    public async apply(input: T, options?: O): Promise<ResultCode<T>> {
        for (const codemod of this.codemods) {
            const result = await codemod.apply(input, options);

            if (result.modified) {
                return result;
            }
        }

        return {
            modified: false,
            result: input,
        };
    }
}
