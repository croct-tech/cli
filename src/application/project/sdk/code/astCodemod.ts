import type {namedTypes as Ast} from 'ast-types';
import {print} from 'recast';
import {Codemod, ResultCode} from '@/application/project/sdk/code/transformation';
import {parse} from '@/application/project/sdk/code/parser';

export class AstCodemod implements Codemod<string> {
    private readonly codemods: Array<Codemod<Ast.Node>>;

    public constructor(...transformers: Array<Codemod<Ast.Node>>) {
        this.codemods = transformers;
    }

    public apply(input: string): ResultCode<string> {
        if (this.codemods.length === 0) {
            return {
                modified: false,
                result: input,
            };
        }

        const ast = parse(input, ['jsx', 'typescript']);

        const result = this.codemods.reduce<ResultCode<Ast.Node>>(
            (next, transformer) => {
                const transformation = transformer.apply(next.result);

                return {
                    modified: transformation.modified || next.modified,
                    result: transformation.result,
                };
            },
            {
                modified: false,
                result: ast,
            },
        );

        if (!result.modified) {
            return {
                modified: false,
                result: input,
            };
        }

        return {
            modified: true,
            result: print(result.result).code,
        };
    }
}
