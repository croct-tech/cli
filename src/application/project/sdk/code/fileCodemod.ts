import {readFile, writeFile} from 'fs/promises';
import {Codemod, CodemodError, ResultCode} from '@/application/project/sdk/code/transformation';
import {formatCause} from '@/application/error';

export class FileCodemod implements Codemod<string> {
    private readonly codemod: Codemod<string>;

    public constructor(codemod: Codemod<string>) {
        this.codemod = codemod;
    }

    public async apply(input: string): Promise<ResultCode<string>> {
        let source = '';

        try {
            source = await readFile(input, 'utf-8');
        } catch (error) {
            if (error.code !== 'ENOENT') {
                throw new CodemodError(`Failed to read file: ${formatCause(error)}`);
            }
        }

        const result = await this.codemod.apply(source);

        if (result.modified) {
            try {
                await writeFile(input, result.result, {encoding: 'utf-8', flag: 'w'});
            } catch (error) {
                throw new CodemodError(`Failed to write file: ${formatCause(error)}`);
            }
        }

        return result;
    }
}
