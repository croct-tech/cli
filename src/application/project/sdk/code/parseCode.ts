import type {namedTypes as Ast} from 'ast-types';
import {print} from 'recast';
import {Codemod, CodemodOptions, ResultCode} from '@/application/project/sdk/code/codemod';
import {Language, parse} from '@/application/project/sdk/code/parser';

export type Configuration<O extends CodemodOptions> = {
    codemod: Codemod<Ast.Node, O>,
    languages: Language[],
};

export class ParseCode<O extends CodemodOptions> implements Codemod<string, O> {
    private readonly codemod: Codemod<Ast.Node, O>;

    private readonly languages: Language[];

    public constructor(configuration: Configuration<O>) {
        this.codemod = configuration.codemod;
        this.languages = configuration.languages;
    }

    public async apply(input: string, options?: O): Promise<ResultCode<string>> {
        const result = await this.codemod.apply(parse(input, this.languages), options);

        if (!result.modified) {
            return {
                modified: false,
                result: input,
            };
        }

        return {
            modified: true,
            result: print(result.result).code,
        };
    }
}
