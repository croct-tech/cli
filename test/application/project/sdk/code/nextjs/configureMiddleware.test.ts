import {resolve} from 'path';
import {MiddlewareOptions, ConfigureMiddleware} from '@/application/project/sdk/code/nextjs/configureMiddleware';
import {loadFixtures} from '../fixtures';
import {ParseCode} from '@/application/project/sdk/code/parseCode';

describe('ConfigureMiddleware', () => {
    const defaultOptions: MiddlewareOptions = {
        import: {
            module: '@croct/plug-next/middleware',
            highOrderFunctionName: 'withCroct',
            middlewareFunctionName: 'middleware',
            matcherName: 'matcher',
            matcherLocalName: 'matcher',
            configName: 'config',
        },
    };

    const scenarios = loadFixtures<MiddlewareOptions>(
        resolve(__dirname, '../fixtures/nextjs-middleware'),
        defaultOptions,
        {
            'matcherAlias.ts': {
                import: {
                    ...defaultOptions.import,
                    matcherLocalName: 'croctMatcher',
                },
            },
        },
    );

    it.each(scenarios)('should correctly transform $name', async ({name, fixture, options}) => {
        const transformer = new ParseCode({
            languages: ['typescript'],
            codemod: new ConfigureMiddleware(options),
        });

        const output = await transformer.apply(fixture);

        expect(output.result).toMatchSnapshot(name);
    });
});
