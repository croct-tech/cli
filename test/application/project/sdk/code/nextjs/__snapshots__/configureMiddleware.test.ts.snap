// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigureMiddleware should correctly transform configAfterDefaultExport.ts: configAfterDefaultExport.ts 1`] = `
"import { withCroct, matcher } from "@croct/plug-next/middleware";
import type { NextRequest } from 'next/server'

export const config = {
    matcher: ['/((?!api|_next/static|_next/image|favicon.ico).*)', matcher],
}

export const middleware = withCroct({
    matcher: config.matcher,

    next: function(request: NextRequest) {
        console.log(request.url);
    }
});
"
`;

exports[`ConfigureMiddleware should correctly transform configAfterMiddlewareWithReference.ts: configAfterMiddlewareWithReference.ts 1`] = `
"import { withCroct, matcher } from "@croct/plug-next/middleware";
import type { NextRequest } from 'next/server'

// pattern
const PATTERN = '/((?!api|_next/static|_next/image|favicon.ico).*)';

// matcher
class Matcher {
    static readonly PATTERN = PATTERN;
}

// getMatcher
function getMatcher() {
    return Matcher.PATTERN;
}

// currentMatcher
const currentMatcher = getMatcher();

// currentConfig
const currentConfig = {
    matcher: [
        ...(Array.isArray(currentMatcher) ? currentMatcher : [currentMatcher]),
        matcher
    ],
}

// middleware
export const middleware = withCroct({
    matcher: currentConfig.matcher,

    next: function(request: NextRequest) {
        console.log(request.url);
    }
});

function unrelated() {
}

// Non-global scope
{
    // Different scope, should not be considered
    const PATTERN = null;

    function getMatcher() {
    }

    class Matcher {
    }
}

// config
export const config = currentConfig;
"
`;

exports[`ConfigureMiddleware should correctly transform configAfterNamedExport.ts: configAfterNamedExport.ts 1`] = `
"import { withCroct, matcher } from "@croct/plug-next/middleware";
import type { NextRequest } from 'next/server'

export const config = {
    matcher: ['/((?!api|_next/static|_next/image|favicon.ico).*)', matcher],
}

export default withCroct({
    matcher: config.matcher,

    next: function middleware(request: NextRequest) {
        console.log(request.url);
    }
});
"
`;

exports[`ConfigureMiddleware should correctly transform configInvalidReference.ts: configInvalidReference.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";
import type { NextRequest } from 'next/server'

export default withCroct(function middleware(request: NextRequest) {
    console.log(request.url);
});

export const config = bar;
"
`;

exports[`ConfigureMiddleware should correctly transform configWithArrayMatcher.ts: configWithArrayMatcher.ts 1`] = `
"import { withCroct, matcher } from "@croct/plug-next/middleware";
import type { NextRequest } from 'next/server'

export const config = {
    matcher: ['/((?!api|_next/static|_next/image|favicon.ico).*)', matcher],
}

export const middleware = withCroct({
    matcher: config.matcher,

    next: function(request: NextRequest) {
        console.log(request.url);
    }
});
"
`;

exports[`ConfigureMiddleware should correctly transform configWithIndirectVariableReference.ts: configWithIndirectVariableReference.ts 1`] = `
"import { withCroct, matcher } from "@croct/plug-next/middleware";
const regex = '/((?!api|_next/static|_next/image|favicon.ico).*)';

const configValue = {
    matcher: [...(Array.isArray(regex) ? regex : [regex]), matcher],
}

const indirectReference = configValue;

export const config = indirectReference;

export const middleware = withCroct({
    matcher: configValue.matcher,

    next: function(request) {
        console.log(request.url);
    }
});
"
`;

exports[`ConfigureMiddleware should correctly transform configWithStringMatcher.ts: configWithStringMatcher.ts 1`] = `
"import { withCroct, matcher } from "@croct/plug-next/middleware";
export const config = {
    matcher: ['/((?!api|_next/static|_next/image|favicon.ico).*)', matcher],
}

export const middleware = withCroct({
    matcher: config.matcher,

    next: function(request) {
        console.log(request.url);
    }
});
"
`;

exports[`ConfigureMiddleware should correctly transform configWithVariableMatcher.ts: configWithVariableMatcher.ts 1`] = `
"import { withCroct, matcher } from "@croct/plug-next/middleware";
const regex = '/((?!api|_next/static|_next/image|favicon.ico).*)';

export const config = {
    matcher: [...(Array.isArray(regex) ? regex : [regex]), matcher],
}

export const middleware = withCroct({
    matcher: config.matcher,

    next: function(request) {
        console.log(request.url);
    }
});
"
`;

exports[`ConfigureMiddleware should correctly transform configWithVariableReference.ts: configWithVariableReference.ts 1`] = `
"import { withCroct, matcher } from "@croct/plug-next/middleware";
const regex = '/((?!api|_next/static|_next/image|favicon.ico).*)';

const configValue = {
    matcher: [...(Array.isArray(regex) ? regex : [regex]), matcher],
}

export const config = configValue;

export const middleware = withCroct({
    matcher: configValue.matcher,

    next: function(request) {
        console.log(request.url);
    }
});
"
`;

exports[`ConfigureMiddleware should correctly transform configWithoutMatcher.ts: configWithoutMatcher.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";
export const config = {
}

export const middleware = withCroct(function(request) {
    console.log(request.url);
});
"
`;

exports[`ConfigureMiddleware should correctly transform defaultExportAnonymousFunction.ts: defaultExportAnonymousFunction.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";

export default withCroct(function() {
    console.log('middleware');
});
"
`;

exports[`ConfigureMiddleware should correctly transform defaultExportArrowFunction.ts: defaultExportArrowFunction.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";

export default withCroct(() => {
    console.log('middleware');
});
"
`;

exports[`ConfigureMiddleware should correctly transform defaultExportArrowFunctionReference.ts: defaultExportArrowFunctionReference.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";
const anything = withCroct((request) => console.log(request.url));

export default anything;
"
`;

exports[`ConfigureMiddleware should correctly transform defaultExportArrowFunctionWithBodyReference.ts: defaultExportArrowFunctionWithBodyReference.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";
const anything = withCroct((request) => {
    console.log(request.url);
})

export default anything;
"
`;

exports[`ConfigureMiddleware should correctly transform defaultExportClass.ts: defaultExportClass.ts 1`] = `
"export default class Middleware {
    // invalid
}
export { middleware, config } from "@croct/plug-next/middleware";
"
`;

exports[`ConfigureMiddleware should correctly transform defaultExportFunctionDeclaration.ts: defaultExportFunctionDeclaration.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";

export default withCroct(function anything(request) {
    console.log(request.url);
});
"
`;

exports[`ConfigureMiddleware should correctly transform defaultExportFunctionDeclarationReference.ts: defaultExportFunctionDeclarationReference.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";

const anything = withCroct(function(request) {
    console.log(request.url);
});

export default anything;
"
`;

exports[`ConfigureMiddleware should correctly transform defaultExportFunctionExpressionReference.ts: defaultExportFunctionExpressionReference.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";
const anything = withCroct(function (request) {
    console.log(request.url);
})

export default anything;
"
`;

exports[`ConfigureMiddleware should correctly transform defaultExportIndirectReference.ts: defaultExportIndirectReference.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";
const anything = function (request) {
    console.log(request.url);
}

const something = withCroct(anything);

export default something;
"
`;

exports[`ConfigureMiddleware should correctly transform empty.ts: empty.ts 1`] = `"export { middleware, config } from "@croct/plug-next/middleware";"`;

exports[`ConfigureMiddleware should correctly transform existingAliasedHofCall.ts: existingAliasedHofCall.ts 1`] = `
"import { withCroct as croctMiddleware } from "@croct/plug-next/middleware";

export default croctMiddleware(function () {
    console.log('middleware');
});
"
`;

exports[`ConfigureMiddleware should correctly transform existingAliasedMiddlewareCall.ts: existingAliasedMiddlewareCall.ts 1`] = `
"import { middleware } from "@croct/plug-next/middleware";

export default middleware(function () {
    console.log('middleware');
});
"
`;

exports[`ConfigureMiddleware should correctly transform existingConfig.ts: existingConfig.ts 1`] = `
"import { withCroct, matcher } from "@croct/plug-next/middleware";
export const config = {
    matcher: ['/((?!api|_next/static|_next/image|favicon.ico).*)', matcher],
}

export default withCroct({
    matcher: config.matcher
});
"
`;

exports[`ConfigureMiddleware should correctly transform existingConfigMatcher.ts: existingConfigMatcher.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";
import { matcher } from "@croct/plug-next/middleware";

export const config = {
    matcher: matcher,
}

export default withCroct({
    matcher: config.matcher
});
"
`;

exports[`ConfigureMiddleware should correctly transform existingConfigReexport.ts: existingConfigReexport.ts 1`] = `
"export { config } from "@croct/plug-next/middleware";
export { middleware } from "@croct/plug-next/middleware";
"
`;

exports[`ConfigureMiddleware should correctly transform existingHofCall.ts: existingHofCall.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";

export default withCroct(function () {
    console.log('middleware');
});
"
`;

exports[`ConfigureMiddleware should correctly transform existingImport.ts: existingImport.ts 1`] = `
"import { withCroct, matcher } from "@croct/plug-next/middleware";

export const config = {
    matcher: ['.*', matcher]
};

export const middleware = withCroct({
    matcher: config.matcher,

    next: function() {
        console.log('middleware');
    }
});
"
`;

exports[`ConfigureMiddleware should correctly transform existingImportAliased.ts: existingImportAliased.ts 1`] = `
"import { withCroct as croctMiddleware, matcher as croctMatcher } from "@croct/plug-next/middleware";

export const config = {
    matcher: ['.*', croctMatcher]
};

export const middleware = croctMiddleware({
    matcher: config.matcher,

    next: function() {
        console.log('middleware');
    }
});
"
`;

exports[`ConfigureMiddleware should correctly transform existingLocalConfigAndMiddlewareReexport.ts: existingLocalConfigAndMiddlewareReexport.ts 1`] = `
"export { middleware } from "@croct/plug-next/middleware";

export const config = {
    matcher: ['/((?!api|_next/static|_next/image|favicon.ico).*)'],
}
"
`;

exports[`ConfigureMiddleware should correctly transform existingMiddlewareAndAliasedConfigMatcher.ts: existingMiddlewareAndAliasedConfigMatcher.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";
import { matcher as croctMatcher } from "@croct/plug-next/middleware";

export const config = {
    matcher: ['/((?!api|_next/static|_next/image|favicon.ico).*)', croctMatcher],
}

export default withCroct({
    matcher: config.matcher,

    next: function() {
        console.log('middleware');
    }
});
"
`;

exports[`ConfigureMiddleware should correctly transform existingMiddlewareAndConfigArrayMatcher.ts: existingMiddlewareAndConfigArrayMatcher.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";
import { matcher } from "@croct/plug-next/middleware";

export const config = {
    matcher: ['/((?!api|_next/static|_next/image|favicon.ico).*)', matcher],
}

export default withCroct({
    matcher: config.matcher,

    next: function() {
        console.log('middleware');
    }
});
"
`;

exports[`ConfigureMiddleware should correctly transform existingMiddlewareAndConfigReexport.ts: existingMiddlewareAndConfigReexport.ts 1`] = `
"export { middleware, config } from "@croct/plug-next/middleware";
"
`;

exports[`ConfigureMiddleware should correctly transform existingMiddlewareCall.ts: existingMiddlewareCall.ts 1`] = `
"import { middleware as croctMiddleware } from "@croct/plug-next/middleware";

export default croctMiddleware(function () {
    console.log('middleware');
});
"
`;

exports[`ConfigureMiddleware should correctly transform existingMiddlewareReexport.ts: existingMiddlewareReexport.ts 1`] = `
"export { middleware } from "@croct/plug-next/middleware";
export { config } from "@croct/plug-next/middleware";
"
`;

exports[`ConfigureMiddleware should correctly transform matcherAlias.ts: matcherAlias.ts 1`] = `
"import { withCroct, matcher as croctMatcher } from "@croct/plug-next/middleware";

export const config = {
    matcher: ['.*', croctMatcher]
};

export const middleware = withCroct({
    matcher: config.matcher,

    next: function() {
        console.log('middleware');
    }
});
"
`;

exports[`ConfigureMiddleware should correctly transform namedExportArrowFunction.ts: namedExportArrowFunction.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";
export const middleware = withCroct(() => console.log('middleware'));
"
`;

exports[`ConfigureMiddleware should correctly transform namedExportArrowFunctionWithBody.ts: namedExportArrowFunctionWithBody.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";
export const middleware = withCroct(() => {
    console.log('middleware');
})
"
`;

exports[`ConfigureMiddleware should correctly transform namedExportFunctionDeclaration.ts: namedExportFunctionDeclaration.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";

export const middleware = withCroct(function() {
    console.log('middleware');
});
"
`;

exports[`ConfigureMiddleware should correctly transform namedExportFunctionExpression.ts: namedExportFunctionExpression.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";
export const middleware = withCroct(function() {
    console.log('middleware');
})
"
`;

exports[`ConfigureMiddleware should correctly transform namedSpecifiedExport.ts: namedSpecifiedExport.ts 1`] = `
"import { withCroct, matcher } from "@croct/plug-next/middleware";

const config = {
    matcher: ['.*', matcher]
};

const middleware = withCroct({
    matcher: config.matcher,

    next: function() {
        console.log('middleware');
    }
})

export { middleware, config };
"
`;

exports[`ConfigureMiddleware should correctly transform specifiedExportWithAliases.ts: specifiedExportWithAliases.ts 1`] = `
"import { withCroct, matcher } from "@croct/plug-next/middleware";
function unrelated() {
}

const _config = {
    matcher: ['.*', matcher]
};

const _middlewareFn = withCroct({
    matcher: _config.matcher,

    next: function() {
        console.log('middleware');
    }
})

export {
    _middlewareFn as default,
    _config as config
};
"
`;

exports[`ConfigureMiddleware should correctly transform unrelatedExports.ts: unrelatedExports.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";
export function foo() {
}

export default withCroct(() => {

});

export function middleware() {
}
"
`;
